In [13]:
import pandas as pd
import numpy as np
import re

#### **VENDORS REGISTRY**

In [110]:
vendors = pd.read_csv('../data/raw/vendors/illinois_state_vendor_registry_as_of_20230614.csv', encoding='cp1252',
                    dtype=str)

vendors = vendors[['Company Name',
 'Address',
 'City',
 'State',
 'Zip',
]]
vendors.head()

Company Name                        Address  \
0               'D' CONSTRUCTION, INC.            1488 SOUTH BROADWAY   
1    #1 Professional Cleaning Services                   2205 Mayfair   
2  *PRADO & RENTERIA CPAS, Prof. Corp.     1837 South Michigan Avenue   
3   @Veterans Construction Group, LLC.                 2608 W 25th St   
4                        0x Global LLC  6177 N Lincoln Ave, Suite 306   

          City State      Zip  
0    COAL CITY    IL  \t60416  
1  Westchester    IL  \t60154  
2      Chicago    IL  \t60616  
3      Chicago    IL  \t60608  
4      Chicago    IL  \t60659

In [111]:
#Check for missing values in company name
print('Missing number of vendors:', vendors[vendors['Company Name'].isna()].shape[0])
#Only keep rows that have a company name
vendors = vendors[~vendors['Company Name'].isna()].reset_index(drop=True)

Missing number of vendors: 2


In [112]:
#Cleaning the company name: removing numbers, punctuation and special characters
vendors['clean_name_vendor'] = np.nan

for i in range(vendors.shape[0]):
    vendors.loc[i, 'clean_name_vendor'] = re.sub('[^A-Za-z]+', '', vendors.loc[i, 'Company Name']).lower()

In [114]:
vendors[['Company Name', 'clean_name_vendor']]

Company Name             clean_name_vendor
0                  'D' CONSTRUCTION, INC.              dconstructioninc
1       #1 Professional Cleaning Services  professionalcleaningservices
2     *PRADO & RENTERIA CPAS, Prof. Corp.     pradorenteriacpasprofcorp
3      @Veterans Construction Group, LLC.  veteransconstructiongroupllc
4                           0x Global LLC                    xgloballlc
...                                   ...                           ...
3538                     ZOI Incorporated               zoiincorporated
3539             Zonatherm Products, Inc.          zonathermproductsinc
3540                     Zuber Lawler LLP                zuberlawlerllp
3541        Zuriel Medical Staffing, Inc.      zurielmedicalstaffinginc
3542    Generated from the B2Gnow System.   generatedfromthebgnowsystem

[3543 rows x 2 columns]

#### **AWARDED CONTRACTS**

In [124]:
contracts = pd.read_csv('../data/raw/contracts/illinois_bidbuy_contracts_as_of_20230614.csv')
contracts.head()

Contract / Blanket #  Bid Solicitation #  \
0               CIB6354390                 NaN   
1               CIB5736630                 NaN   
2  23-563WCC-WCC56-P-39787                 NaN   
3  23-563WCC-WCC56-P-38456                 NaN   
4  23-551HDA-LEGAL-P-39047                 NaN   

                              Description                      Vendor  \
0                      Prescription Drugs  CaremarkPCS Health, L.L.C.   
1     ConnectYourCare - Flexible Spending        ConnectYourCare, LLC   
2  Sole Source- WCRI CompScope Reports 23                        WCRI   
3       EDI Hosting and Maintenance FY 23                  Ebix, Inc.   
4   Foley & Lardner Bond Counsel (Exempt)         FOLEY & LARDNER LLP   

  Type Code Dollars Spent to Date  \
0         B                 $0.00   
1         A                 $0.00   
2         D                 $0.00   
3         C                 $0.00   
4         Z                 $0.00   

                                      Organization      Status  Begin Date  \
0                CMS - Central Management Services  3PS - Sent  07/01/2015   
1                CMS - Central Management Services  3PS - Sent  07/01/2019   
2  WCC - Illinois Worker's Compensation Commission  3PS - Sent  08/15/2022   
3  WCC - Illinois Worker's Compensation Commission  3PS - Sent  07/01/2022   
4              HDA - Housing Development Authority  3PS - Sent  07/07/2022   

     End Date  
0  06/30/2024  
1  06/30/2023  
2  06/30/2023  
3  06/30/2023  
4  07/06/2024

In [125]:
#Check for missing values in vendors
print('Missing number of vendors:', contracts[contracts['Vendor'].isna()].shape[0])

Missing number of vendors: 0


In [126]:
#Cleaning the company name: removing numbers, punctuation and special characters
contracts['clean_name_contr'] = np.nan

for i in range(contracts.shape[0]):
    contracts.loc[i, 'clean_name_contr'] = re.sub('[^A-Za-z]+', '', contracts.loc[i, 'Vendor']).lower()

In [127]:
contracts[['Vendor', 'clean_name_contr']]

Vendor  \
0                      CaremarkPCS Health, L.L.C.   
1                            ConnectYourCare, LLC   
2                                            WCRI   
3                                      Ebix, Inc.   
4                             FOLEY & LARDNER LLP   
...                                           ...   
1371                                Midland Paper   
1372                                     WEX BANK   
1373   Change Healthcare Pharmacy Solutions, Inc.   
1374       Cognizant Technology Solutions US Corp   
1375  International Business Machines Corporation   

                              clean_name_contr  
0                         caremarkpcshealthllc  
1                           connectyourcarellc  
2                                         wcri  
3                                      ebixinc  
4                              foleylardnerllp  
...                                        ...  
1371                              midlandpaper  
1372                                   wexbank  
1373      changehealthcarepharmacysolutionsinc  
1374        cognizanttechnologysolutionsuscorp  
1375  internationalbusinessmachinescorporation  

[1376 rows x 2 columns]

#### **MERGING BOTH DATASETS**

In [130]:
contract_vendor = pd.merge(contracts, vendors, left_on='clean_name_contr', right_on='clean_name_vendor', how='left')
contract_vendor.head(5)

Contract / Blanket #  Bid Solicitation #  \
0               CIB6354390                 NaN   
1               CIB5736630                 NaN   
2  23-563WCC-WCC56-P-39787                 NaN   
3  23-563WCC-WCC56-P-38456                 NaN   
4  23-551HDA-LEGAL-P-39047                 NaN   

                              Description                      Vendor  \
0                      Prescription Drugs  CaremarkPCS Health, L.L.C.   
1     ConnectYourCare - Flexible Spending        ConnectYourCare, LLC   
2  Sole Source- WCRI CompScope Reports 23                        WCRI   
3       EDI Hosting and Maintenance FY 23                  Ebix, Inc.   
4   Foley & Lardner Bond Counsel (Exempt)         FOLEY & LARDNER LLP   

  Type Code Dollars Spent to Date  \
0         B                 $0.00   
1         A                 $0.00   
2         D                 $0.00   
3         C                 $0.00   
4         Z                 $0.00   

                                      Organization      Status  Begin Date  \
0                CMS - Central Management Services  3PS - Sent  07/01/2015   
1                CMS - Central Management Services  3PS - Sent  07/01/2019   
2  WCC - Illinois Worker's Compensation Commission  3PS - Sent  08/15/2022   
3  WCC - Illinois Worker's Compensation Commission  3PS - Sent  07/01/2022   
4              HDA - Housing Development Authority  3PS - Sent  07/07/2022   

     End Date      clean_name_contr          Company Name  \
0  06/30/2024  caremarkpcshealthllc                   NaN   
1  06/30/2023    connectyourcarellc                   NaN   
2  06/30/2023                  wcri                   NaN   
3  06/30/2023               ebixinc                   NaN   
4  07/06/2024       foleylardnerllp  FOLEY & LARDNER, LLP   

                              Address     City State      Zip  \
0                                 NaN      NaN   NaN      NaN   
1                                 NaN      NaN   NaN      NaN   
2                                 NaN      NaN   NaN      NaN   
3                                 NaN      NaN   NaN      NaN   
4  321 North Clark Street, Suite 2800  CHICAGO    IL  \t60654   

  clean_name_vendor  
0               NaN  
1               NaN  
2               NaN  
3               NaN  
4   foleylardnerllp

In [131]:
contract_vendor.shape

(1376, 17)

In [132]:
contract_vendor.isna().sum()

Contract / Blanket #        0
Bid Solicitation #       1376
Description                 0
Vendor                      0
Type Code                   0
Dollars Spent to Date       0
Organization                0
Status                      0
Begin Date                  0
End Date                    0
clean_name_contr            0
Company Name              857
Address                   857
City                      857
State                     857
Zip                       857
clean_name_vendor         857
dtype: int64

Note that of the 1376 contracts awarded, there are 857 that were not awarded to vendors in the registry.

#### **BEP vendors**

In [133]:
bep = pd.read_csv("../data/clean/bep_company_addresses_all.csv")
bep.head()

Company Name Address Type  \
0  #1 Professional Cleaning Services      Mailing   
1  #1 Professional Cleaning Services     Physical   
2                     #9 Design, LLC      Mailing   
3                     #9 Design, LLC     Physical   
4           'Bouche' Enterprises LLC      Mailing   

                        Street         City State  ZIP Code  
0                P.O. Box 7555  Westchester    IL     60154  
1                 2205 Mayfair  Westchester    IL     60154  
2   28 East Saint Charles Road   Villa Park    IL     60181  
3  524 West Saint Charles Road   Villa Park    IL     60181  
4    3511 North Osceola Avenue      Chicago    IL     60634

Note that some company names are duplicated in the data since they have both a mailing and physical address. Since we want to locate the vendors we will keep only the Physical addresses.

In [135]:
#Filtering physical address
bep = bep.loc[bep['Address Type'] == 'Physical'].reset_index(drop=True)

In [136]:
#Cleaning the company name: removing numbers, punctuation and special characters
bep['clean_name_bep'] = np.nan

for i in range(bep.shape[0]):
    bep.loc[i, 'clean_name_bep'] = re.sub('[^A-Za-z]+', '', bep.loc[i, 'Company Name']).lower()

In [137]:
bep[['Company Name', 'clean_name_bep']]

Company Name  \
0            #1 Professional Cleaning Services   
1                               #9 Design, LLC   
2                     'Bouche' Enterprises LLC   
3     'Merica Made Painting & Decorating Corp.   
4                       *VDR Development Group   
...                                        ...   
4972                         mk communications   
4973                        myWHY Agency, Inc.   
4974                               netlogx LLC   
4975                           r6catalyst, LLC   
4976                   rSTAR Technologies, LLC   

                        clean_name_bep  
0         professionalcleaningservices  
1                            designllc  
2                 boucheenterprisesllc  
3     mericamadepaintingdecoratingcorp  
4                  vdrdevelopmentgroup  
...                                ...  
4972                  mkcommunications  
4973                    mywhyagencyinc  
4974                        netlogxllc  
4975                      rcatalystllc  
4976              rstartechnologiesllc  

[4977 rows x 2 columns]

#### **MERGING BEP DATA WITH CONTRACTS AND VENDORS**

In [139]:
contr_vendor_bep = pd.merge(contract_vendor, bep, left_on='clean_name_contr', right_on='clean_name_bep', how='left')

In [141]:
contr_vendor_bep[contr_vendor_bep['clean_name_bep'].isna()]

Contract / Blanket #  Bid Solicitation #  \
0                  CIB6354390                 NaN   
1                  CIB5736630                 NaN   
2     23-563WCC-WCC56-P-39787                 NaN   
3     23-563WCC-WCC56-P-38456                 NaN   
4     23-551HDA-LEGAL-P-39047                 NaN   
...                       ...                 ...   
1371     17-416CMS-BOSS4-P-99                 NaN   
1372     17-416CMS-BOSS4-P-39                 NaN   
1373               14Z0879001                 NaN   
1374               13Z1819001                 NaN   
1375               000MY14222                 NaN   

                                   Description  \
0                           Prescription Drugs   
1          ConnectYourCare - Flexible Spending   
2       Sole Source- WCRI CompScope Reports 23   
3            EDI Hosting and Maintenance FY 23   
4        Foley & Lardner Bond Counsel (Exempt)   
...                                        ...   
1371       JPMC PAPER AND ENVELOPES SMALL QTYS   
1372               MC Fleet Fuel Card Services   
1373      Pharmacy Benefits Management (PBMS)    
1374  HFS MMIS Upgrade Project Expert Services   
1375   SBE: Student Information System Support   

                                           Vendor Type Code  \
0                      CaremarkPCS Health, L.L.C.         B   
1                            ConnectYourCare, LLC         A   
2                                            WCRI         D   
3                                      Ebix, Inc.         C   
4                             FOLEY & LARDNER LLP         Z   
...                                           ...       ...   
1371                                Midland Paper         A   
1372                                     WEX BANK         B   
1373   Change Healthcare Pharmacy Solutions, Inc.         B   
1374       Cognizant Technology Solutions US Corp         B   
1375  International Business Machines Corporation         B   

     Dollars Spent to Date                                     Organization  \
0                    $0.00                CMS - Central Management Services   
1                    $0.00                CMS - Central Management Services   
2                    $0.00  WCC - Illinois Worker's Compensation Commission   
3                    $0.00  WCC - Illinois Worker's Compensation Commission   
4                    $0.00              HDA - Housing Development Authority   
...                    ...                                              ...   
1371         $4,502,328.93                CMS - Central Management Services   
1372           $279,800.00                CMS - Central Management Services   
1373                 $0.00             HFS - Healthcare and Family Services   
1374                 $0.00             HFS - Healthcare and Family Services   
1375                 $0.00                   SBE - State Board of Education   

          Status  Begin Date    End Date  ... State_x      Zip  \
0     3PS - Sent  07/01/2015  06/30/2024  ...     NaN      NaN   
1     3PS - Sent  07/01/2019  06/30/2023  ...     NaN      NaN   
2     3PS - Sent  08/15/2022  06/30/2023  ...     NaN      NaN   
3     3PS - Sent  07/01/2022  06/30/2023  ...     NaN      NaN   
4     3PS - Sent  07/07/2022  07/06/2024  ...      IL  \t60654   
...          ...         ...         ...  ...     ...      ...   
1371  3PS - Sent  03/06/2017  07/31/2023  ...     NaN      NaN   
1372  3PS - Sent  07/01/2016  06/30/2025  ...     NaN      NaN   
1373  3PS - Sent  03/17/2014  03/16/2024  ...     NaN      NaN   
1374  3PS - Sent  06/20/2013  06/19/2023  ...     NaN      NaN   
1375  3PS - Sent  07/01/2013  06/30/2023  ...     NaN      NaN   

     clean_name_vendor Company Name_y Address Type Street City_y State_y  \
0                  NaN            NaN          NaN    NaN    NaN     NaN   
1                  NaN            NaN          NaN    NaN    NaN     NaN   
2                  NaN            NaN       

Note: This are the 1238 Non-bep contracts